### ISSUE TITLE
Feature request: Use SubViewport to render the 3D models 

### ISSUE THREAD

**Issue by sc0ttj** [sci-comp/SceneBuilder#6]
This plugin looks amazing! Great work.

I do have a suggestion, if you don't mind.

It should be possible to avoid having to create screenshots for all your assets - and render the 3D model in the UI instead. This has two benefits:

- no need to create your own thumbnail images
- can zoom/spin the item on hovering over the asset in the GUI

Maybe using a scene tree similar to this one:

![godot-3d-in-ui--using-subviewport](https://github.com/sci-comp/SceneBuilder/assets/2726610/7c3c9192-dccc-4b31-9afa-f916951b51d9)

(The relevant bit is `TextureButton -> SubViewportContainer -> SubViewport -> Camera3D`)

And maybe some code like so:

```
# Grab the camera3D inside your GUI

# Load and insert the model, if not added before
if camera_3d.get_child_count() == 0:
    # Get instance of the 3D scene to load, and add to the current scene
    var my_3d_scene = load(FOO.get_scene_file_path()).instantiate()
    
    # Add the scene into the viewport
    camera_3d.add_child(my_3d_scene)
    
    # zoom out and center
    my_3d_scene.global_translate(Vector3(0.0, -0.25, -1.5)) 
```

If this is not possible, undesired, or outside the scope of this project, then by all means close this issue.

Thanks.

**Comment by sci-comp** [comment]
What a clever idea! I'd really love to clear our all those icons from the Data folder-- I feel like Godot already has a method for this somewhere, since MeshLibrary resources already create icons for us.

The only concern I have with your suggestion would be related to performance. Users may wish to have very large collections, so I worry about the performance cost of having 100-1000 cameras in the scene.

Camera performance probably scales with polygons rendered, plus some overhead? Maybe cameras are more efficient than I realize?

What do you think?

**Comment by sc0ttj** [comment]
Thanks for the reply.

For low poly models, I would imagine performance should not be _too_ bad, until 100s of models are listed. 

Though that is purely based on my laptop being a 10 year old potato, onboard Intel gfx, and that as far as I can tell, it's only really particles and complex collision shapes that slow things down for me.

But I suppose it mostly depends on the hardware being used, the complexity of the models, and how many models users put into a single category.

...It _might_ be worth investigating using a Camera3D to auto-save the thumbnail image when adding the item (users could even choose the view angle, then click "save thumbnail").  And then when browsing the items you could use that screenshot until the user hovers over the item, at which point you load the 3d scene proper, let it spin.

Another option may be to try some form of pagination - where you enable/disable the camera view as items scroll in/out of view.. though I've no idea how easy/viable this would be when working with Godots GUI/panels.

Yet another option for performance improvements may be to load in multiple models in a grid, all shown with one camera, make each item clickable with area3d or similar. Essentially you would have fewer camera3Ds, and each one you have would show a grid of items.

All just suggestions of course.

Thanks.

**Comment by sci-comp** [comment]
Well, I guess I'll leave it like it is for now. I referenced a few different tools, and they all seem to simply create low quality thumbnails somewhere. It's simple, and we've already got it working. Might as well keep it as it is.

Although, while looking at this today, I noticed some unrelated bugs in `create_scene_builder_items.gd`, shown in the commit below.

Thank you this feature request! It's really encouraging to have others people looking at (and possibly using) this project!
